Add correlation function

asizemore opened the pull request:
Resolves #33 
Requires https://github.com/VEuPathDB/veupathUtils/pull/20

Adds a core correlation function that accepts data tables. Then there are other functions that accept abundance data and metadata that just do formatting really and pass the data to the core corr computation (say that 5 times fast!). Added spearman and pearson methods because we got both for free.

This PR also adds a `SampleMetadata` class

To Dos

- [x] make core corr function
- [x] make SampleMetadata class
- [x] make corr fn that accepts samplemetadata and abundance data
- [x] change stats names
- [x] test stats names
- [x] document
- [x] add checks to data.table x data.table fn

I started making the toJSON changes, but realized that will require work in veupathutils. I think it'd be better if i made that a separate PR. This one is getting large anyways.

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,100 @@
+    # Check that incoming metadata meets requirements
+    if (!'data.table' %in% class(sampleMetadata)) {
+      data.table::setDT(sampleMetadata)
+    }
```
are these not validation the relevant classes can perform? like can AbundanceData not enforce that data and sampleMetadata are of class data.table?

asizemore replied:
they can enforce that class by erring, but they cant' change the class if it's not correct. So yeah we could totally just move all this validation to the class! That'd clean up all the methods i think.

asizemore replied:
#46

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,100 @@
+
+    # Replace NA values with 0
+    if (naToZero) {
+      veupathUtils::setNaToZero(df)
```
im starting to wonder if this should be a method of AbundanceData rather than called by each of its methods individually

asizemore replied:
i could see the different methods having opinions about setting na to zero or not. Like maybe one correlation method wants to use it but another doesn't. I'll learn more when we actually add the more interesting correlation methods

d-callan replied:
imputing zeroes seems to be about appropriately modeling the data, not about the compute? not sure i understand

d-callan replied:
so we probably could have the class do it automatically, but im not sure thats what we want. i could see keeping track of the original/ unmodified data being worthwhile. i think what id prefer is an accessor fxn for the AbundanceData class. we should probably be writing those anyhow, but this maybe increases the pressure there.. the idea i think would be to write a method for AbundanceData like `getDataTable` or something, which by default imputes zeroes if necessary, but can be overridden if desired.

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,100 @@
+      # Reorder sampleMetadata so that it matches the order in df
+      sampleMetadata <- sampleMetadata[match(df[[recordIdColumn]],sampleMetadata[[recordIdColumn]]), ]
+      veupathUtils::logWithTime("Reordered sampleMetadata rows based on abundance data sample id order.", verbose)
+    }
```
this also feels like validation that should be done by AbundanceData?

asizemore replied:
#46

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,100 @@
+    ## Find numeric metadata
+    numericMetadataCols <- veupathUtils::findNumericCols(sampleMetadata)
+    if (length(numericMetadataCols) < 1) {
+      stop("correlation requires at least one continuous metadata variable.")
```
we should be careful here. were checking for numeric and then claiming we need continuous. dates are also continuous and some numbers are discrete.. id prefer, if we can, to use our variable/ collections metadata to identify whether we have appropriate data for this..

asizemore replied:
good call. Updated in recent commits

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,100 @@
+    statistics <- data.frame(
+      var1 = meltedCorrResult[['rn']],
+      var2 = meltedCorrResult[['variable']],
+      correlationCoef = meltedCorrResult[['value']]
```
can we be more explicit than var1 and var2?

asizemore replied:
i'd be happy to but i dont have a much better idea. The network uses terms like 'source' and 'target' but that sort of directionality isn't appropriate here. I also thought about 'correlate1' and 'correlate2' but that is hard to distinguish from the verb "correlate" lol and also seems like i'm already assuming these two variables are correlated. 
We could at least say "variable1" and "variable2" though?

d-callan replied:
when we add support for metadata vs metadata and assay vs assay, do we see multiple correlation methods w different signatures? is it possible to explicitly say one of these are abundance data and the other sample metadata?

asizemore replied:
that's a good question. Yes i expect different signatures so definitely possible. The other question i have though (thinking aloud here) is does having the abundance and metadata names here in the stats df help us or make some code further down the line harder because it has to account for lots of different stats col names? Like then the api would have to account for different col names. Totally possible, but then in the ds it'll call out these columns explicitly in order to do some network things and though it'd be easier to have it all say the same var1, var2, it'll be easier to read and probably to validate if they have real names. Hmmmm by the end of this thought train i was at the opposite opinion of where i started.

d-callan replied:
fwiw i also dont like calling it var1 when its values are variables according to EDA verbiage..

asizemore replied:
this approach will make it hard to validate there are only two variables and then the correlation results. So some stats object with assay, metadata, and another assay could get through. Also when we do correlation assay v assay we'll end up with the same problem and they'll be named assay1 and assay2. I could keep it variable1 and variable2 and then add another property that says var1 is an assay and var2 is metadata... hmmm

asizemore commented:
a note to my future self:
we're going to have a lot of different ways to use the correlation function. Some with metadata, sometimes with two assays, sometimes with one, sometmes with random genomics data, or just metadata. My current idea for how this should grow is to have one basic correlation function that takes exactly two data frames (or matrices whatevs) and does whatever correlation method. Then we have lots of correlation methods that do different validations and data wrangling based on the inputs, then eventually make two data.frames and send them to the basic correlation function. That way we're not crazy repeating code and can appropriately handle what to do with lots o' different inputs

asizemore commented:
Update before i switch my brain to diff abund
- added support for dates
- added support for real variable metadata
- not totally convinced on what to do about the var1 var2 thing. The elements of these columns are variables so maybe the column name should be correlate1Name correlate2Name? There's also a discussion around about naming them based on taxa or assay or something that's coming in. That has to do also with breaking up the plugin. Worth some more thinking before deciding.
- should shore up how handling the `variables` input within the function. Should it be part of the signature? Is `NULL` okay for the default or should it be an empty `VariableMetadataList`? Probably the latter...

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,173 @@
+
+    ## Format results
+    # Rename the correlation results columns so that they are a little more friendly
+    names(corrResult)[names(corrResult) == "data1"] <- data1Name
```
not sure i like this. wondering if we can have a CorrelationResult class or something, that has a slot for the data.frame your default method returns, but also slots for things like 'data1Metadata` and `data2Metadata` where we could say that data1 is sample data or assay data etc.

d-callan replied:
also, why are these params of the method? doesnt the signature here guarantee were working w assay v metadata? am i misunderstanding what the names are for?

asizemore replied:
i understood the goal to be to let us more specifically access the data, so if we only know the classes we'll run into trouble when we have assay v assay. In that case we'd have to basically name it assay1 and assay2 if the method just had to define these names itself. So I was thinking we'd pass in the collection ID to dataName1 and "metadata" or whatever into dataName2. 

But i think i probably misunderstood the original use for this type of thing. Because even if we have a CorrelationResult class then for assay v assay  it'd also have "Assay" in both dataYMetadata slots?

d-callan replied:
I think the goal ultimately is to know what data1 and data2 represent. We want to be able to reference them generically I think, so the component+ viz for ex can just say put whatever 'data1' is over here or there. But also it needs to know what data1 fundamentally IS to do things like put display labels places? And/ or validate data1 is what its meant to be too maybe? Maybe the metadata slot is a VariableMetadata object?

asizemore replied:
got it. Okay that makes sense!

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,173 @@
+      }
+
+    } else {
+      warning("No variable metadata supplied. Using all numeric columns of sampleMetadata for correlation.")
```
i dont like that the behavior is different whether you pass specific variables or not. presumably the default should also include dates, if were going to support them when requested?

d-callan replied:
also will this mean that we treat differently a var w type = number and shape = ordinal depending if variables is present?

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    ## Extract appropriate metadata columns
+    if (!is.null(variables)) {
```
im ok w this, but i dont think its a lot of work to prune a data.table to only the cols youre interested in, if you want to avoid the `variables` param. like if you wanted to just assume SampleMetadata includes only the things we care about, particularly for this first iteration, id be ok. the reason i offer, is bc im not sure if i like this vs a method on SampleMetadata like `prune` which takes a `VariableMetadataList`..  and maybe we dont have to decide right now?

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,173 @@
+    ## Extract appropriate metadata columns
+    if (!is.null(variables)) {
+      # Use the data shape to find appropriate columns in the metadata
+      inputMetadataCols <- veupathUtils::findColNamesByPredicate(variables, function(x) {identical(x@dataShape@value, "CONTINUOUS")})
```
this silently ignores requested columns? maybe the user doesnt realize categoricals dont make sense here.. they wont get any feedback unless all cols are removed.

d-callan commented on R/method-correlation.R:
```diff
@@ -0,0 +1,173 @@
+    result <- new("ComputeResult")
+    result@name <- 'correlation'
+    result@recordIdColumn <- recordIdColumn1
+    result@ancestorIdColumns <- ancestorIdColumns1
```
are we making an assumption here that the AbundanceData is on a child node of the SampleMetadata? or can we somehow explicitly validate or compare the ancestor ids of the two data.tables?

asizemore replied:
I need the record ids to match so the sample metadata should be already wrangled into the assay entity space. So i guess we could check to make sure the ancestor ids match? We should add a similar check to the diff abund method, then, too

d-callan reviewed (commented):
big improvement :)

asizemore commented:
updated based on the feedback. Removed `variables` entirely, which will help us get things moving. In the future we need to revisit the conversation so that we can optimally handle dates. I expect i'll be back here in a month anyways so we can do that then after we learn a bit more about how it all works together.

d-callan reviewed (commented):
its been a while since i looked at this last, so maybe ive just lost the thread a bit, but im confused. if you can explain what youre trying to do w the new metadata slots, thatd be helpful. also we may need to think more about the signature for the correlation method(s)..

asizemore commented on R/method-correlation.R:
```diff
@@ -0,0 +1,125 @@
+#' @param method string defining the type of correlation to run. The currently supported values are 'spearman' and 'pearson'
+#' @param verbose boolean indicating if timed logging is desired
+#' @return ComputeResult object
+setMethod("correlation", signature("data.table", "data.table"), function(data1, data2, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
```
my plan is to add methods to this function that support compositional data. The AbundanceData would probably need an isCompositional flag and then if we try to run a non-compositional method on compositional data it could err. Or there could be a compositional subclass and it gets a method. Either way i think the way it's set up will make adding the compositional methods pretty "simple"!

d-callan replied:
Are we simply not worried about it for Dec? I'd guess we need it for release though no? Since compositional data are critical here..

asizemore commented on R/method-correlation.R:
```diff
@@ -0,0 +1,125 @@
+  if (!is.null(data1@sampleMetadata)) {
+    warning("Replacing sampleMetadata in data1 with data2 sampleMetadata")
+  }
+  data1@sampleMetadata <- data2
```
Using the SampleMetadata helps us be explicit about what type of correlation we're running. Let's say the correlation only used an AbundanceData object. Are we running taxa x taxa? Taxa x metadata? The metadata x metadata? This line feels odd i agree but it's the easiest way to run all of the appropriate checks on the metadata to make sure it works wiht the taxa abundances. I can make that more clear with some comments.

d-callan replied:
I still don't like it.. what if the method was `correlateAbundancesAndMetadata` or something and took only the AbundanceData object? That's also clear and the API is cleaner. The metadata v metadata one could take a single object SampleMetadata as well no?

d-callan replied:
I just feel like the generic data v data fxn is more an internal helper and these others should be methods of classes properly. And their API should make sense from that perspective and from the perspective of someone trying to use them rather than dev them.

d-callan commented on R/class-CorrelationComputeResult.R:
```diff
@@ -0,0 +1,45 @@
+    parameters = NA_character_,
+    data1Metadata = NA_character_,
+    data2Metadata = NA_character_
+), validity = check_compute_result)
```
Should be using the new validation fxn?

d-callan reviewed (approved):
I suppose im willing to assume we'll figure out the compositional thing. I would be happier knowing we had an answer ready bc it seems central, but I'll try to trust our future selves.. the API still makes me uncomfortable though, like it was built by a dev for a dev. 

Ill approve, so you can keep on, but I'm nervous....
